Use docker-machine and docker-compose for image creation and instance provisioning

fgimenez opened the pull request.

elopio commented on README.md:
```diff
@@ -9,13 +9,35 @@ There are also jobs configured for accessing the Snappy Product Integration envi
 ## Provision
 
-There are provision scripts for creating the CI environment locally and in the cloud.
+There are provision scripts for creating the CI environment locally and in the cloud. These scripts use docker-machine and docker-compose, so you need to have them installed, see [https://docs.docker.com/machine/install-machine/](here) and [https://docs.docker.com/compose/install/](here). Once the provision is finished you can use the common commands of machine and compose, for instance, to list  ...
```
Should we avoid the ones from the ubuntu archive?

fgimenez replied:
docker-machine isn't yet in the archive, and the version of docker-compose is old for the format that we are using here.

elopio commented on README.md:
```diff
@@ -9,13 +9,35 @@ There are also jobs configured for accessing the Snappy Product Integration envi
+
+    eval $(docker-machine env snappy-jenkins-remote)
+
+Once thats done you can use your local docker client as if you were on the remote machine, try `docker info`, `docker images`, `docker ps` and the like.
```
that's

elopio replied:
I think it should be: in the remote machine.
But I'm not sure.

fgimenez replied:
done

elopio commented on README.md:
````diff
@@ -35,7 +57,7 @@ In this case you need to have OpenStack credentials loaded, for example by sourc
     $ source path/to/openstack/credentials/novarc
 
-Before Executing the provision script you should have at least the ```$OS_USERNAME``` and ```$OS_REGION_NAME``` environment variables set. This credentials are used for creating the hosts where the CI environment itself is going to run, and maybe different to the ones used later for the Ubuntu Core instances, more on this later.
+Before Executing the provision script you should have at least the ```$OS_USERNAME``` and ```$OS_REGION_NAME``` environment variables set. This credentials are used for creating the hosts where the CI environment itself is going to run, and maybe different to the ones used later for the Ubuntu Core instances, more on this later. The scripts also assume that you have a valid private key at `~/.can ...
````
s/Executing/executing

elopio replied:
s/maybe/may be

fgimenez replied:
done

elopio commented on bin/cloud-provision.sh:
```diff
@@ -16,6 +16,53 @@ fi
 NOVARC_PATH=$1
 SPI_CREDENTIALS_PATH=$2
 OPENSTACK_CREDENTIALS_DIR="$JENKINS_HOME/.openstack"
+PRIVATE_KEY_PATH=~/.canonistack/${OS_USERNAME}_${OS_REGION_NAME}.key
```
Why isn't this one an argument?
I have my personal keys, the team keys for canonistack and scalingstack. It would be nicer as an argument.

fgimenez replied:
Ok, added as last argument to the create and provision script, with de default value being `~/.canonistack/${OS_USERNAME}_${OS_REGION_NAME}.key`

elopio commented on bin/cloud-provision.sh:
```diff
@@ -16,6 +16,53 @@ fi
+    local INSTANCE_ID=$1
+    local retry=30
+    local INSTANCE_IP=""
+    INSTANCE_IP=$(openstack server show $INSTANCE_ID | grep 'addresses' | awk '{print $4}' | cut -d= -f2)
```
You can pass -c to select a column. I was using something like this, but I'm sure you can prettify it with awk and remove the grep pipe:
ip=$(openstack server show -c addresses -f shell $INSTANCE_ID | cut -f3 -d = | tr -d '"')

elopio replied:
oh, I've just found there's a -f value. We won't need any pipes.

fgimenez replied:
done, I can remove the grep and awk pipes, cut is still needed because the `addresses` field looks like `net_ues-snappy-integration=10.42.63.60`

elopio commented on bin/cloud-provision.sh:
```diff
@@ -16,6 +16,53 @@ fi
+launch_instance(){
+    local IMAGE_NAME=$1
+
+    IMAGE_ID=$(openstack image list | grep "$IMAGE_NAME" | head -1 | awk '{print $4}')
```
openstack image list -c Name -f value

fgimenez replied:
In fact we don't need the image id, the name is enough to launch the instance. I've used the `-c` and `-f` options later to get the instance ip.

elopio commented on bin/cloud-provision.sh:
```diff
@@ -32,18 +79,30 @@ copy_credentials() {
-send_and_execute "$INSTANCE_IP" "$JENKINS_HOME" "./bin/remote/redeploy.sh"
+add_instance_to_docker_machine "$NAME_REMOTE" "$INSTANCE_IP" "$PRIVATE_KEY_PATH"
+
+sleep 5
```
A comment for this sleep would be nice.

fgimenez replied:
leftover, removed

elopio commented on bin/common.sh:
```diff
@@ -1,215 +1,23 @@
     openstack security group delete $SECGROUP
     openstack security group create --description "snappy-jenkins secgroup" $SECGROUP
-    # ports 22 and 8080 only accessible from the vpn, port 8081
+    # ports 22, 8080 and 2376 only accessible from the vpn, port 8081
```
Missing a comment about the reason for this port 2376

fgimenez replied:
done

elopio commented on bin/create-image.sh:
```diff
@@ -4,21 +4,9 @@ set -x
-}
+PRIVATE_KEY_PATH=${1:-~/.canonistack/${OS_USERNAME}_${OS_REGION_NAME}.key}
+KEYPAIR_NAME=${OS_USERNAME}_${OS_REGION_NAME}
+IMAGE_NAME="uci/cloudimg/$DIST-amd64.img"
```
What's uci here? Whose images are we consuming?

fgimenez replied:
In order to reduce the amount of shell script needed to determine the name of the image to be used (the daily ones carry a timestamp) I was looking for images with fixed names and found this ones, they are (or were) used by the Ubuntu CI Engine team for running tests before "landing new code in the distro" (from the project page). See [1] and [2] 

[1] http://bazaar.launchpad.net/~canonical-ci-engineering/uci-engine/trunk/view/head:/britney_proxy/README.images#L38
[2] http://bazaar.launchpad.net/~canonical-ci-engineering/uci-engine/trunk/view/head:/ci-utils/ci_utils/image_store/__init__.py#L95

elopio commented on bin/create-image.sh:
```diff
@@ -36,26 +24,42 @@ create_snapshot(){
 
-INSTANCE_ID=$(launch_instance "$IMAGE_NAME")
+    openstack keypair delete $keypair_name
+    ssh-keygen -y -f $private_key > $public_key
```
We might have a problem here. For running the autopkgtests we need the private key of the scalingstack keypair to be in the .ssh/config of the slaves. This private key is not the same as the ~/.ssh/id_rsa we are using, right?

fgimenez replied:
No, this one is used for accessing the host(s) where the containers run, the one you mention should be handled with the rest of secrets.

elopio commented on bin/create-image.sh:
```diff
@@ -36,26 +24,42 @@ create_snapshot(){
 
-exit
+IP=$(docker-machine ip "$NAME_REMOTE")
+ID=$(openstack server list | grep "$IP" | awk '{ print $2 }')
```
Is that $NAME_REMOTE the name of the server?
If so, you could do openstack server show -c id -f value "$NAME_REMOTE"

fgimenez replied:
No, it is `$NAME-remote`, that is `snappy-jenkins-remote`, it's set in the scripts. The reason for having both remote and local names is because from the same docker-machine cli we are going to be able to manage both environments and we need to distinguish them by name.

elopio commented on config/compose/cluster.yml:
```diff
@@ -0,0 +1,39 @@
+
+  jenkins-slave-xenial-3:
+    extends: jenkins-slave-xenial-1
+    command: -username admin -password snappy -executors 2 -name jenkins-slave-xenial-3 -labels xenial -master http://jenkins-master-service:8080
```
What's this -username? I thought that was going to be the user that runs the jenkins jobs.
We have a problem because currently the jobs are being run as root. Looking at the docs I thought I could use -username jenkins-slave, but now I'm confused.

fgimenez replied:
According to [1] that's the Jenkins username for authentication, not related to the underlying OS user running the jobs AFAIK. I've added a `user` option to the compose definition of the slave containers, let me know if that works for you.

[1] https://wiki.jenkins-ci.org/display/JENKINS/Swarm+Plugin#SwarmPlugin-AvailableOptions

elopio commented:
This looks great to me. I tried the local provision but failed with:
Package python3.4 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'python3.4' has no installation candidate
ERROR: Service 'jenkins-slave-xenial-1' failed to build: The command '/bin/sh -c apt-get update && apt-get install --no-install-recommends -qy   build-essential   ca-certificates   grub-common   python-junitxml   sudo subunit   snappy-cloud-image   snappy-tests-job   squashfs-tools    ubuntu-core-security-utils   bzr dpkg-dev git mercurial pyflakes python3.4   python3-apt python3-docopt python3-coverage   python3-fixtures python3-jsonschema python3-mccabe   python3-pep8 python3-pip python3-requests   python3-testscenarios python3-testtools python3-yaml   python3-lxml && rm -rf /var/lib/apt/lists/*' returned a non-zero code: 100

fgimenez commented:
> I tried the local provision but failed

Yes, it seems that python3.4 is no longer available in xenial [1], not sure of the reason for that. It seems that previous layers install python3.5, could we drop the python3.4 dependency and work with the newest version?

[1] https://launchpad.net/ubuntu/xenial/+source/python3.4

fgimenez commented:
Thanks a lot for the review, I think that all comments are addressed, I've also removed the python3.4 dependency from the xenial slave and it builds ok, let me know if the jobs work as expected.

I've also added a new section in the README regarding redeploys, with compose they are much more straightforward and quick. We should use them when we have container changes, no need to spin up new instances, just stopping, pulling the changes and bringing the cluster up again.

elopio commented:
I had a successful local deploy. Trying the cloud now.
Can you please add instructions to tear everything down? I tried docker-machine stop and kill, and the machine is still there when I docker-machine ls.

elopio commented:
I got this error on create-image: Creating compose_jenkins-slave-xenial-3_1
ERROR: An HTTP request took too long to complete. Retry with --verbose to obtain debug information.

It didn't fail the execution, so now I'm not sure if the prev image is good. I guess I should delete the new one, and rename the prev.

fgimenez commented:
I've added check for the exit status of docker-compose so that if something goes wrong the images wouldn't be renamed and eventually removed. I've retriggered the image creation in lcy01 and all seems to be good, please take a look.

elopio commented:
I got the same error:
Status: Downloaded newer image for ubuntucore/snappy-jenkins-slave-xenial:latest
Creating compose_jenkins-slave-xenial-1_1
ERROR: An HTTP request took too long to complete. Retry with --verbose to obtain debug information.
If you encounter this issue regularly because of slow network conditions, consider setting COMPOSE_HTTP_TIMEOUT to a higher value (current value: 60).

So it might be good to follow the suggestion and raise the timeout.

fgimenez commented:
Ok done, I've set it to 180

elopio commented:
Deployed to http://10.55.32.66:8080/ \o/
landing it. Tomorrow I can do a sync to refresh the main instance.
